src/rust/cryptography-x509-validation/src/lib.rs

```diff
@@ -11,6 +11,289 @@ pub mod policy;
+            ));
+        }
+
+        // Look in the store's root set to see if the working cert is listed.
+        // If it is, we've reached the end.
+        //
+        // Observe that no issuer connection or signature verification happens
+        // here: inclusion in the root set implies a trust relationship,
+        // even if the working certificate is an EE or intermediate CA.
+        if self.store.contains(working_cert) {
+            let mut constraints = vec![];
+            self.build_name_constraints(&mut constraints, extensions)?;
+            return Ok((vec![working_cert.clone()], constraints));
+        }
+
+        // Otherwise, we collect a list of potential issuers for this cert,
+        // and continue with the first that verifies.
+        for issuing_cert_candidate in self.potential_issuers(working_cert) {
+            // A candidate issuer is said to verify if it both
+            // signs for the working certificate and conforms to the
+            // policy.
+            let issuer_extensions = issuing_cert_candidate.extensions()?;
+            if let Ok(next_depth) = self.policy.valid_issuer(
+                issuing_cert_candidate,
+                working_cert,
+                current_depth,
+                &issuer_extensions,
+            ) {
+                let result = self.build_chain_inner(
+                    issuing_cert_candidate,
+                    next_depth,
+                    false,
+                    &issuer_extensions,
+                );
+                if let Ok(result) = result {
+                    let (remaining, mut constraints) = result;
+                    // Name constraints are not applied to self-issued certificates unless they're
+                    // the leaf certificate in the chain.
+                    //
+                    // NOTE: We can't simply check the `current_depth` since self-issued
+                    // certificates don't increase the working depth.
+                    let skip_name_constraints = cert_is_self_issued(working_cert) && !is_leaf;
+                    if skip_name_constraints
+                        || self
+                            .apply_name_constraints(&constraints, extensions)
+                            .is_ok()
+                    {
+                        let mut chain: Vec<Certificate<'chain>> = vec![working_cert.clone()];
+                        chain.extend(remaining);
+                        self.build_name_constraints(&mut constraints, extensions)?;
+                        return Ok((chain, constraints));
+                    }
+                }
+            }
+        }
+
+        // We only reach this if we fail to hit our base case above, or if
+        // a chain building step fails to find a next valid certificate.
+        Err(ValidationError::CandidatesExhausted)
+    }
+
+    fn build_chain(&self, leaf: &'a Certificate<'chain>) -> Result<Chain<'chain>, ValidationError> {
```

Comment:
IIUC, this will build a chain that is valid per the policy or fail entirely. OpenSSL had an issue where because it performed enforcement of certain properties **after** building a candidate chain it could end up building one that it would later reject and then there was no mechanism to build a path that would be accepted. That issue is irrelevant here because we enforce validity of all properties as we build the chain. Except I'm unclear how it would handle the following:

What happens if you're building a chain `EE -> IntB -> IntA`, but `IntA` adds a constraint that blocks the `EE`. However, `IntC` was also present and would have allowed a valid path to be built? (e.g. `EE -> IntC -> root`)